### PR TITLE
BABEL-4411 set transaction isolation cmd change (2)

### PR DESCRIPTION
**Pull request by tanscorpio7**
### Description
Currently babelfish transactions start with session default isolation level. Users can not change this isolation level after transaction has started. On running the set transaction isolation inside a transaction, current transaction isolation does not change but all subsequent transaction will pick up this new isolation level.

Changes in this PR:

Allow changing the current transaction isolation level if no DML query has been run inside transaction.
Abort the transaction if set transaction isolation level is run after a DML query in a transaction block unless the new isolation level is same as current isolation level.
Introduce an escape hatch to ignore any invalid set transaction isolation level stmt so that transaction does not abort.

#### Extension Changes:
1. Introduce escape hatch for set transaction isolation
2. New attribute in PLtsql_stmt_execsql structure. This attribute is needed to identify if the stmt is set transaction isolation during initialisation of internal savepoint and if set, we skip the internal save point for the current stmt.
This needs to be done since postgres does not allow setting isolation level inside a save point.
3. Create a new internal function which is responsible for handling all the checks before setting isolation level GUCs. This internal function should be directly called from the grammer tree.

### Issues Resolved

BABEL-4411

### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c [hunk 1]
```diff
@@ -1765,3 +1765,47 @@ TsqlForClauseSubselect(Node *selectstmt)
...
+	return 0;
+}
+
+static void
+bbf_set_tran_isolation(char *new_isolation_level_str)
+{
+	const char*		ignore_isolation;
+	const int 		new_isolation_int_val = isolation_to_int(new_isolation_level_str);
+
+	if( new_isolation_int_val != DefaultXactIsoLevel )
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c]: if (abc), not if( abc )

File: test/JDBC/input/transactions/BABEL_4411.sql [hunk 2]
```diff
@@ -0,0 +1,212 @@
+-- Escape hatch disabled
+
+SET TRANSACTION ISOLATION LEVEL READ COMMITTED
+GO
+SELECT current_setting('transaction_isolation')
+SELECT current_setting('default_transaction_isolation')
+SELECT set_config('babelfishpg_tsql.escape_hatch_set_transaction_isolation_level', 'strict', false)
```
Comment (surendravishnoi) [inline test/JDBC/input/transactions/BABEL_4411.sql]: can we set this escape hatch using TSQL set syntax? if yes, please use TSQL syntax

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c [hunk 3]
```diff
@@ -1765,3 +1765,47 @@ TsqlForClauseSubselect(Node *selectstmt)
...
+	{
+		if( FirstSnapshotSet || IsSubTransaction() || 
+				( new_isolation_int_val == XACT_SERIALIZABLE && RecoveryInProgress() ))
+		{
+			ignore_isolation = GetConfigOption("babelfishpg_tsql.escape_hatch_set_transaction_isolation_level",
+												false, false);
+
+			if(strcmp(ignore_isolation, "ignore"))
+				elog(ERROR, "SET TRANSACTION ISOLATION failed, transaction aborted, set escape hatch "
+					"'babelfishpg_tsql.escape_hatch_set_transaction_isolation_level' to ignore such error");
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c]: for TSQL syntax, do not use babelfishpg_tsql prefix in error message

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y [hunk 4]
```diff
@@ -4226,8 +4222,7 @@ tsql_OptTranName:
 tsql_IsolationLevel:
 			ISOLATION LEVEL tsql_IsolationLevelStr
 				{
-					$$ = list_make1(makeDefElem("transaction_isolation",
-									makeStringConst($3, @3), @1));
+					bbf_set_tran_isolation($3);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y]: I do not think it is ideal to do the real change as part of parsing. Let us have an internal C function can call set isolation level from there

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 5]
```diff
@@ -3406,6 +3409,28 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
...
+				VariableSetStmt *variable_set = (VariableSetStmt *) parsetree;
+
+				if(strcmp(variable_set->name, "SESSION CHARACTERISTICS") == 0)
+				{
+					ListCell   		*head;
+
+					foreach(head, variable_set->args)
+					{
+						DefElem		*item = (DefElem *) lfirst(head);
+						A_Const		*con = (A_Const *) item->arg;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: instead of con, isolation_level

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 6]
```diff
@@ -5583,3 +5608,47 @@ bbf_ExecDropStmt(DropStmt *stmt)
...
+{
+	const char*		ignore_isolation;
+	const int 		new_isolation_int_val = isolation_to_int(new_isolation_level_str);
+
+	if(new_isolation_int_val != DefaultXactIsoLevel)
+	{
+		if(FirstSnapshotSet || IsSubTransaction() || 
+				(new_isolation_int_val == XACT_SERIALIZABLE && RecoveryInProgress()))
+		{
+			ignore_isolation = GetConfigOption("babelfishpg_tsql.escape_hatch_set_transaction_isolation_level",
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: see if we can directly use variable escape_hatch_set_transaction_isolation_level

File: test/JDBC/expected/BABEL_4411_2.out [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+read committed
+~~END~~
+
+~~START~~
+text
+read committed
+~~END~~
+
+
+BEGIN TRANSACTION
```
Comment (surendravishnoi) [inline test/JDBC/expected/BABEL_4411_2.out:35]: can we also have tests without begin transaction?

**Review by surendravishnoi — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]

**Pull request merged**